### PR TITLE
Support for user-defined Eq impls

### PR DESCRIPTION
**Pull request by tkadur**
Related to #1549; user-defined equality is a prerequisite for user-defined comparison.

### CODE REVIEW

**Review by zygoloid — COMMENTED** [review]
I had been intending a somewhat different implementation strategy: instead of providing an intrinsic for each of the various builtin operations, I'd been intending to not rewrite them at all when the operand types are integer / floating point types known by explorer. Then the interpreter can continue to have a simple implementation of various operator expressions and we get a more efficient and simpler representation for the primitive operations.

We'd still need `impl`s in the prelude, but they can use the primitive forms via operator notation:
```
impl i32 as Eq {
  fn Op[me: Self](rhs: Self) -> Bool { return me == rhs; }
}
```

What do you think?

**Comment by tkadur** [comment]
I wrote up that version for comparison in https://github.com/tkadur/carbon-lang/pull/1. I'm not sure it resulted in a significant decrease in complexity (and as implemented it loses the free implicit conversion of the RHS for int/string equality), but I'll defer to your choice on this.

File: explorer/testdata/struct/empty.carbon
Comment (jonmeow) [inline explorer/testdata/struct/empty.carbon]: Can you please move this and related constructs into the prelude as well?

**Review by jonmeow — COMMENTED** [review]
This looks great, thank you! Just one small things, I think it makes sense here to keep the various Eq implementations in prelude.carbon.

**Comment by jonmeow** [comment]
@zygoloid My leaning here is to use this kind of approach for consistency; it does make traces more complex, but I think it also is closer to the approach we want out of the toolchain, so it seems good to mirror more closely.

File: explorer/testdata/struct/empty.carbon
Comment (tkadur) [reply explorer/testdata/struct/empty.carbon]: So there currently appears to be an issue with `impl` resolution on struct types (see https://carbon.compiler-explorer.com/z/r9q8T9bPd). I originally moved this out of the prelude as a workaround for that.

**Comment by zygoloid** [comment]
> @zygoloid My leaning here is to use this kind of approach for consistency; it does make traces more complex, but I think it also is closer to the approach we want out of the toolchain, so it seems good to mirror more closely.

I disagree: I think it's important in the toolchain that we don't desugar primitive operations like integer addition into real function calls that will need to be inlined. Doing that desugaring and inlining will create a lot of additional compile-time cost that we should be working to avoid, and risks producing worse code especially at `-O0`. In explorer I think we could go either way, but creating a large collection of named intrinsic functions doesn't seem obviously better to me than having predefined semantics for certain operators and will mean that explorer and the toolchain diverge in their prelude implementation.

I think I'd be OK with having an intrinsic function (however we spell those -- explorer is doing something that has not gone through design review here) and defining the `impl` with an alias:

```
impl i32 as AddWith(i32) where .Result == i32  {
  alias Op = __intrinsic_int_add;
}
```

... but I don't think the explorer infrastructure is ready for that yet. (We'd need intrinsics to be more like actual functions, rather than having intrinsic calls be a special kind of expression.)

Until / unless we do something like that, I think special-casing the primitive operations will be more in-line with what we'd do in the toolchain.

**Comment by zygoloid** [comment]
Possibly a better option than having both a predeclared intrinsic function and an alias would be to make `i32.(AddableWith(i32).Op)` be the intrinsic function, perhaps by declaring it with an attribute:
```
impl i32 as AddWith(i32) where .Result == i32  {
  @Builtin("int.add")
  fn Op[me: i32](rhs: i32) -> i32;
}
```
This would avoid the need to have predefined `__`-prefixed identifiers at all, which seems pretty clean.

**Comment by jonmeow** [comment]
@zygoloid What we're doing right now in cases of intrinsics is a thin wrapper, e.g.:

```
class Heap {
  fn New[T:! Type, me: Self](x : T) -> T* {
    return __intrinsic_new(x);
  }
  fn Delete[T:! Type, me: Self](p : T*) {
    __intrinsic_delete(p);
  }
}
```

I agree in the toolchain we probably wouldn't want the function indirection. However, I also think it would be weird to structure this as intrinsic `Heap_New` and `Heap_Delete` calls, rather than providing some sort of library-based interface.

Both of your examples seem better to me than the current state. But if you're saying this would be a good solution:

```
impl i32 as AddWith(i32) where .Result == i32  {
  alias Op = __intrinsic_int_add;
}
```

 I just view this as a useful step on the way to it:

```
impl i32 as AddWith(i32) where .Result == i32  {
  fn Op[me: i32](rhs: i32) -> i32 {
    return __intrinsic_int_add(me, rhs);
  }
}
```

The key difference in my mind is that this approach pushes code into `__intrinsic_int_add`, where the ideal state simply needs some reshaping done on `__intrinsic_int_add`. We would removing the special-casing of `+` where the interface isn't supported.

In keeping with `+`, it avoids the `AddWith` interface, but instead requires a special-casing of `+` to understand that some types don't use the `AddWith` interface. This is what I'm saying I feel diverges from the language design more.

I'm wondering if this would be good to discuss further in a better channel, like VC or such?

**Comment by jonmeow** [comment]
Just to be sure -- I think this will affect what happens as `i*`, `u*`, and `f*` are added. Right now the special-casing code is fairly limited, but either it'll need to expand or we'll be adding more intrinsics. This is kind of a fork in options right now.

**Comment by jonmeow** [comment]
@zygoloid and I spoke about this, and we have #1804 as a tracking issue. I think what's being done here is in keeping with what we discussed as an interim solution, but before I approve I just want to get a final confirmation from him.

File: explorer/interpreter/builtins.cpp
Comment (zygoloid) [inline explorer/interpreter/builtins.cpp]: This should not be necessary any more: the above code should handle `Eq` automatically.

File: explorer/data/prelude.carbon
Comment (zygoloid) [inline explorer/data/prelude.carbon]: Please can you define an `EqWith` following https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/comparison_operators.md#equality ? We don't support `constraint` yet, so I don't think we can define `Eq` in terms of it, but this should avoid the need to change the `==` implementation when that works.

File: explorer/interpreter/builtins.h
Comment (zygoloid) [inline explorer/interpreter/builtins.h]: As per my other comment, please use `EqWith` here.

File: explorer/interpreter/type_checker.cpp
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp]: I think it's probably clearer to a reader of the diagnostic to refer to these as "argument 1" and "argument 2", not 0 and 1.

**Comment by zygoloid** [comment]
> @zygoloid and I spoke about this, and we have #1804 as a tracking issue. I think what's being done here is in keeping with what we discussed as an interim solution, but before I approve I just want to get a final confirmation from him.

Yes, I think this is in line with what we discussed.

File: explorer/interpreter/type_checker.cpp
Comment (tkadur) [reply explorer/interpreter/type_checker.cpp]: Done.

File: explorer/interpreter/builtins.h
Comment (tkadur) [reply explorer/interpreter/builtins.h]: Done.

File: explorer/data/prelude.carbon
Comment (tkadur) [reply explorer/data/prelude.carbon]: Done.

File: explorer/interpreter/builtins.cpp
Comment (tkadur) [reply explorer/interpreter/builtins.cpp]: Done.

**Comment by jonmeow** [comment]
Note with the latest commit:
//explorer/testdata:struct/empty.carbon.test                             FAILED in 0.3s

**Comment by jonmeow** [comment]
Sorry, now there's a resolve conflict. :-/

**Review by jonmeow — APPROVED** [review]
Thank you for carrying this PR through! I think it's in good shape, and I'm going to merge it now.

**Pull request merged**
